### PR TITLE
feat(ingest): add urn modification helper

### PR DESCRIPTION
**Pull request by hsheth2**
These helper methods will be useful for building an URN lowercasing transformer and an URN modification source "auto_*" processor.

This PR makes some tweaks to the codegen in order to drive URN discovery using the metadata model and annotations.

## Checklist

- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/scripts/avro_codegen.py [hunk 1]
```diff
@@ -66,29 +65,77 @@ def load_schemas(schemas_path: str) -> Dict[str, dict]:
...
+    ] = {}  # schema name -> list of (field name, type)
+
+    # First, we need to load the PDL files and find all urn arrays.
+    for pdl_file in Path(pdl_path).glob("**/*.pdl"):
+        pdl_text = pdl_file.read_text()
+        arrays = re.findall(
+            r"^\s*(\w+)\s*:\s*(?:optional\s+)?array\[(\w*Urn)\]",
+            pdl_text,
+            re.MULTILINE,
+        )
```
Comment (asikowitz) [inline metadata-ingestion/scripts/avro_codegen.py:80]: Feel like I'm missing context on this file and when it's used, but this looks like it relies on all our `Urn` schemas (? what's the term here?) having the `Urn` suffix. I guess that's pretty logical, but have we documented this anywhere? At the least, can you add a comment onto the relevant function, cause regexes are hard to parse

File: metadata-ingestion/scripts/avro_codegen.py [hunk 2]
```diff
@@ -66,29 +65,77 @@ def load_schemas(schemas_path: str) -> Dict[str, dict]:
...
+        if isinstance(nested, (avro.schema.EnumSchema, avro.schema.FixedSchema)):
+            continue
+        assert isinstance(nested, avro.schema.RecordSchema)
+
+        # Patch normal urn types.
+        field: avro.schema.Field
+        for field in nested.fields:
+            java_class = field.props.get("java", {}).get("class")
+            if java_class:
+                if java_class.startswith("com.linkedin.pegasus2avro.common.urn."):
```
Comment (asikowitz) [inline metadata-ingestion/scripts/avro_codegen.py:108]: Could be on one line

File: metadata-ingestion/scripts/avro_codegen.py [hunk 3]
```diff
@@ -66,29 +65,77 @@ def load_schemas(schemas_path: str) -> Dict[str, dict]:
...
-    with unittest.mock.patch("avro.schema.Names.add_name", add_name):
-        cleaned_schema = avro.schema.make_avsc_object(merged)
 
-    # Convert back to an Avro schema JSON representation.
-    class MappingProxyEncoder(json.JSONEncoder):
-        def default(self, obj):
-            if isinstance(obj, types.MappingProxyType):
-                return dict(obj)
-            return json.JSONEncoder.default(self, obj)
+def patch_schema(schema: dict, urn_arrays: Dict[str, List[Tuple[str, str]]]) -> dict:
```
Comment (asikowitz) [inline metadata-ingestion/scripts/avro_codegen.py:93]: So this is just adding the `Urn` prop to all schema fields that are urns, and `urn_is_array`?

**Review by asikowitz — COMMENTED** [review]
Tried to review this but honestly don't understand what's going on here

File: metadata-ingestion/scripts/avro_codegen.py (hunk 2 above)
Comment (hsheth2) [reply metadata-ingestion/scripts/avro_codegen.py]: fixed

File: metadata-ingestion/scripts/avro_codegen.py (hunk 1 above)
Comment (hsheth2) [reply metadata-ingestion/scripts/avro_codegen.py:85]: added a comment about the assumption

File: metadata-ingestion/scripts/avro_codegen.py (hunk 3 above)
Comment (hsheth2) [reply metadata-ingestion/scripts/avro_codegen.py:98]: yep - I added a comment that clarifies exactly what it does

File: metadata-ingestion/build.gradle [hunk 4]
```diff
@@ -18,7 +18,7 @@ task checkPythonVersion(type: Exec) {
 task environmentSetup(type: Exec, dependsOn: checkPythonVersion) {
   inputs.file file('setup.py')
   outputs.dir("${venv_name}")
-  commandLine 'bash', '-c', "${python_executable} -m venv ${venv_name} && ${venv_name}/bin/python -m pip install --upgrade pip wheel setuptools>=63.0.0"
+  commandLine 'bash', '-c', "${python_executable} -m venv ${venv_name} && ${venv_name}/bin/python -m pip install --upgrade pip wheel 'setuptools>=63.0.0'"
```
Comment (anshbansal) [inline metadata-ingestion/build.gradle:21]: you may need to merge master. This was done https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/build.gradle#LL21C3-L21C3 already

File: metadata-ingestion/tests/unit/serde/test_urn_iterator.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+
+def test_list_urns_upstream():
+    upstream_table = Upstream(
+        dataset=_datasetUrn("upstream_table_1"),
+        type=DatasetLineageTypeClass.TRANSFORMED,
+    )
+
+    urns = list_urns_with_path(upstream_table)
+    assert urns == [
```
Comment (anshbansal) [inline metadata-ingestion/tests/unit/serde/test_urn_iterator.py:28]: can you explain this

File: metadata-ingestion/tests/unit/serde/test_urn_iterator.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+                upstreamType=FineGrainedLineageUpstreamType.DATASET,
+                upstreams=[_datasetUrn("upstream_table_1")],
+                downstreamType=FineGrainedLineageDownstreamType.FIELD,
+                downstreams=[_fldUrn("downstream_table", "c5")],
+            ),
+        ],
+    )
+
+    urns = list_urns_with_path(upstream_lineage)
+    assert urns != [
```
Comment (anshbansal) [inline metadata-ingestion/tests/unit/serde/test_urn_iterator.py:76]: can we add a equal assertion here?

File: metadata-ingestion/src/datahub/utilities/urns/urn_iter.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+from avro.schema import Field, RecordSchema
+
+from datahub.metadata.schema_classes import DictWrapper
+from datahub.utilities.urns.dataset_urn import DatasetUrn
+from datahub.utilities.urns.urn import Urn, guess_entity_type
+
+_Path = List[Union[str, int]]
+
+
+def list_urns_with_path(model: DictWrapper) -> List[Tuple[str, _Path]]:
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/utilities/urns/urn_iter.py:12]: docstring or more tests for this

File: metadata-ingestion/src/datahub/ingestion/graph/client.py [hunk 8]
```diff
@@ -333,15 +331,9 @@ def get_aspects_for_entity(
...
-            record_schema: RecordSchema = aspect_type.__getattribute__(
-                aspect_type, "RECORD_SCHEMA"
-            )
-            if not record_schema:
-                logger.warning(
-                    f"Failed to infer type name of the aspect from the aspect type class {aspect_type}. Continuing, but this will fail."
-                )
-            else:
-                aspect_type_name = record_schema.props["Aspect"]["name"]
+            record_schema = aspect_type.RECORD_SCHEMA
```
Comment (anshbansal) [inline metadata-ingestion/src/datahub/ingestion/graph/client.py:334]: why this change?

**Review by anshbansal — APPROVED** [review]

File: metadata-ingestion/scripts/avro_codegen.py [hunk 9]
```diff
@@ -66,29 +65,89 @@ def load_schemas(schemas_path: str) -> Dict[str, dict]:
     return schemas
 
 
-def merge_schemas(schemas_obj: List[Any]) -> str:
-    # Combine schemas.
-    merged = ["null"] + schemas_obj
+def patch_schemas(schemas: Dict[str, dict], pdl_path: Path) -> Dict[str, dict]:
```
[the remark this answers is not in the document]
Comment (hsheth2) [reply metadata-ingestion/scripts/avro_codegen.py:68]: This is part of the codegen so we can't unit test it directly, but we do have tests that verify the results of this

File: metadata-ingestion/src/datahub/ingestion/graph/client.py (hunk 8 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/ingestion/graph/client.py:334]: Just general code cleanup not that RECORD_SCHEMA is part of the DictWrapper interface

File: metadata-ingestion/tests/unit/serde/test_urn_iterator.py (hunk 5 above)
Comment (hsheth2) [reply metadata-ingestion/tests/unit/serde/test_urn_iterator.py:28]: added a comment

File: metadata-ingestion/tests/unit/serde/test_urn_iterator.py (hunk 6 above)
Comment (hsheth2) [reply metadata-ingestion/tests/unit/serde/test_urn_iterator.py]: fixed - thanks for catching this

File: metadata-ingestion/src/datahub/utilities/urns/urn_iter.py (hunk 7 above)
Comment (hsheth2) [reply metadata-ingestion/src/datahub/utilities/urns/urn_iter.py:12]: added a doc, it has plenty of tests
